### DISCUSSION TITLE
Switch context for specific reconciler

### DISCUSSION THREAD

**Discussion by dev-matil** [operator-framework/java-operator-sdk#1307] [Ideas]
Hello,
I would like to know if it would be reasonable and worth a contribution to allow a controller to set up not only a namespace but also a context so one operator can listen to multiple clusters. 

Would you be interested to discuss this?

**Comment by csviri** [comment]
Hi @dev-matil ,

Fun fact: about 3 years ago when we started the project we started it with multi cluster support, later removed it ( very early stages) :)

But would be interested in your use case. Isn't deploying the operator to every cluster solves your use case?

 It can be not trivial to manage service accouts ( and related tokens ) cross cluster.

**Reply by bs-matil** [reply]
We have a large instalment of high available services which a spread about multiple clusters in multiple regions. The operator (currently mix of go and scripts). Manages the CRDs in a central control cluster and creates the native resources in the "runtime" clusters. 

There are several things which are done on the central level:

- shifting workload between clusters due to failover events
- shifting workload to new clusters in cluster rescheduling events (yes we do this regularly cluster are ephemeral to us) 
- shifting workload between clusters for global blue green rollouts
- ordered updates between multiple clusters (stateful workloads)

**Reply by bs-matil** [reply]
You could compare it from an architectural standpoint to https://cluster-api.sigs.k8s.io/.
We have cluster of clusters eventually.

**Reply by bs-matil** [reply]
Due to our internal skill sets we want to move to a java based stack. So here we are ;)

**Reply by csviri** [reply]
right, sounds reasonable. 

One think to clarify, the relevant namespaces are replicas on those clusters? So the assumption is that, if we have a configuration for on namespaces on a cluster that will be the same on all other "runtime" clusters?

Or the assumption is that this is always cluster scoped operator?

**Reply by csviri** [reply]
Do I understand correctly that basically what we want to achieve here, is have controllers listening on events from multiple clusters, and do reconciliation for all resources from all clusters (of a certain kind of course) ?

**Reply by metacosm** [reply]
That's my understanding as well.

**Reply by bs-matil** [reply]
yes the latter. the namespaces not necessary relate. The control cluster has different ones. The "runtime" clusters will replicated the same namespaces for each cluster added. 

Example.
 ```

 Control Cluster
┌──────────────────────────────┐  Cluster A
│                              │ ┌──────────────────────────────────────────┐
│    Namespace-crds            │ │  Namespace-A            Namespace-B      │
│    ┌──────────────┐      ┌───┼─┤►┌─────────────┐       ┌─────────────┐    │
│    │              │      │   │ │ │             │       │             │    │
│    │ CRD-A        │      │   │ │ │ Deployment  │       │  Deployment │    │
│    │ CRD-B        │      │   │ │ │             │       │             │    │
│    │              │      │   │ │ └─────────────┘       └─────────────┘    │
│    └─────▲────────┘      │   │ │                                          │
│          │               │   │ └──────────────────────────────────────────┘
│          │ Reconcile     │   │
│          │               │   │  Cluster B
│     ┌────┴────────┬──────┘   │  ┌──────────────────────────────────────────┐
│     │             │          │  │  Namespace-A            Namespace-B      │
│     │  Operator   ├──────────┼──┤►┌─────────────┐       ┌─────────────┐    │
│     │             │  Create/ │  │ │             │       │             │    │
│     └─────────────┘  Reconcile  │ │ Deployment  │       │  Deployment │    │
│                              │  │ │             │       │             │    │
│                              │  │ └─────────────┘       └─────────────┘    │
│                              │  │                                          │
└──────────────────────────────┘  └──────────────────────────────────────────┘
```

**Reply by csviri** [reply]
make sense,

First of all my personal opinion is that such feature makes sense, on the other hand brings complexity to the core of the  framework. I won't even consider this very exotic TBH.

This is also a good topic to discuss on a weekly meeting (maybe today?).  @bs-matil  if you could join would be perfect (zoom link on the home github page, code to join is 77777 )

So I'm up to creating a prototype, and see how it looks, what are the problem that we possibly not see at this moment, and see how much complexity it brings on the table. But not sure about the timeline, but if you would be able to work on this, with our support that might be an option too.

@bs-matil Could you also create an issue for this please?

**Comment by bs-matil** [comment]
This is now a Ticket #1308
